### PR TITLE
feat: Add extension for ActiveRecord transactions

### PR DESCRIPTION
**Pull request by tiev**
## Overview
The conventions of this gem have been adopted in many Rails applications using dry-monad/Do-notation. Therefore, it's useful to have an extension for ActiveRecord.

## Details
This PR introduces a new extension to Dry::Operation that allows operations to be wrapped in ActiveRecord transactions. This is useful for ensuring atomicity of operations that involve multiple steps that need to be rolled back in case of failure. The extension provides a `transaction` method that can be used to wrap steps in a transaction. The transaction will be rolled back if any of the steps return a `Dry::Monads::Result::Failure`.

The extension also supports specifying a custom ActiveRecord class to initiate the transaction, which is useful when working with multiple databases.

### CODE REVIEW

**Comment by tiev** [comment]
@waiting-for-dev Hello, I'm actively implementing operations in my apps and am eager to integrate this gem. This feature would be useful for me. Can you help review this PR?

File: lib/dry/operation/extensions/active_record.rb
Comment (waiting-for-dev) [inline lib/dry/operation/extensions/active_record.rb:55]: It's always nice to see how "rom-rb" works, but we probably don't need it in this context 😉 We should add instead a link to AR docs?

File: lib/dry/operation/extensions/active_record.rb
Comment (waiting-for-dev) [inline lib/dry/operation/extensions/active_record.rb:61]:
```suggestion
        #  Wrap the given steps in an ActiveRecord transaction.
```

File: lib/dry/operation/extensions/active_record.rb
Comment (waiting-for-dev) [inline lib/dry/operation/extensions/active_record.rb:91]: If I'm not wrong, `requires_new` is not the default behaviour in AR. Could it be surprising, or maybe could we make it configurable like with the default connexion? Thoughts on this one?

File: spec/integration/extensions/active_record_spec.rb
Comment (waiting-for-dev) [inline spec/integration/extensions/active_record_spec.rb:15]: Although in memory, do we need an `after :all` removing the table? Thinking about leaving a clean slate for future tests creating the same table.

File: spec/integration/extensions/active_record_spec.rb
Comment (waiting-for-dev) [inline spec/integration/extensions/active_record_spec.rb:8]:
```suggestion
  let(:model) do
```

I'm not very fond of `let!` with bang. Is it going to work without it?

File: spec/integration/extensions/active_record_spec.rb
Comment (waiting-for-dev) [inline spec/integration/extensions/active_record_spec.rb:49]:
```suggestion
    instance = Class.new(base) do
     attr_reader :model
     
      def initialize(model)
        @model = model
      end
      
      def call
        transaction do
          step create_record
          step failure
        end
      end

      def create_record
        Success(model.create(bar: "bar"))
      end

      def failure
        Failure(:failure)
      end
    end.new(model)
```

In this way we can get rid of the `descendants` magic, which from experience it may be brittle (maybe that was the need for `let!` with bang above?)

**Review by waiting-for-dev — COMMENTED** [review]
Thanks @tiev, this is looking great and I'm willing to add it to gem (I already had it in the roadmap). Leaving some comments/suggestions, and also pinging @timriley for his view on this one.

File: lib/dry/operation/extensions/active_record.rb
Comment (tiev) [reply lib/dry/operation/extensions/active_record.rb:91]: You're right. It's not the default and should be configurable.
Should this helper method forward options to the Rails transaction method? ([doc](https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html#method-i-transaction), [available options](https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/DatabaseStatements.html#method-i-transaction))
```ruby
            class_exec(@connection) do |default_connection|
              klass.define_method(:transaction) do |connection = default_connection, **options, &steps|
                connection.transaction(**options) do
                  intercepting_failure(-> { raise ::ActiveRecord::Rollback }, &steps)
                end
              end
            end
```

Comment (tiev) [reply lib/dry/operation/extensions/active_record.rb:91]: Besides, should the interfaces from the 2 extensions be consistent? Wanna hear your ideas

Comment (waiting-for-dev) [reply lib/dry/operation/extensions/active_record.rb:91]: Yes, that would be great. No need to enforce consistency between both extensions. I'd say that's a nice to have but not a must

Comment (tiev) [reply lib/dry/operation/extensions/active_record.rb]: I've updated following your feedback.
I updated the interface for transaction options. It also allows default options when including the extension.

**Review by waiting-for-dev — APPROVED** [review]
Thanks for your work on this one, @tiev. Waiting for @timriley feedback before merging.

File: lib/dry/operation/extensions/active_record.rb
Comment (waiting-for-dev) [reply lib/dry/operation/extensions/active_record.rb]: Thanks!

**Comment by timriley** [comment]
@waiting-for-dev If you're happy with this, I'm happy with this 👍🏼

**Comment by timriley** [comment]
Just noting the build is red; we should fix that before merging :)

**Comment by tiev** [comment]
> Just noting the build is red; we should fix that before merging :)

I fixed it and ran `bundle exec rake` successfully on local.

**Comment by waiting-for-dev** [comment]
Hey @tiev, still read :/

**Comment by tiev** [comment]
> Hey @tiev, still read :/

yeah, yardoc warning concludes as failed. I've fixed it and ensured 0 exit code.

**Comment by waiting-for-dev** [comment]
Thanks for your brilliant work on this one, @tiev 🙌

**Pull request merged**
